### PR TITLE
Make SPIFFE ID and TrustDomain compliant with the specs

### PR DESCRIPTION
**Pull request by maxlambrecht**
This PR fixes and adds validations according to the standard: https://github.com/spiffe/spiffe/blob/master/standards/SPIFFE-ID.md

* Add validation SPIFFE ID maximum length of 2048 
* Add validation TrustDomain maximum length of 255
* Fix parse SPIFFE ID to preserve path case.

The tests for SPIFFE ID and TrustDomain were refactored. 

Signed-off-by: Max Lambrecht <maxlambrecht@gmail.com>

### CODE REVIEW

File: java-spiffe-core/src/main/java/io/spiffe/spiffeid/TrustDomain.java [hunk 1]
```diff
@@ -67,10 +69,23 @@ public String toString() {
         return name;
     }
 
+    /**
+     * Returns the trust domain as SPIFFE ID string (e.g. 'spiffe://example.org')
+     *
+     * @return a String formatted as a SPIFFE ID
+     */
+    public String toIdString() {
+        return String.format("%s://%s", SpiffeId.SPIFFE_SCHEME, name);
```
Comment (rturner3) [inline java-spiffe-core/src/main/java/io/spiffe/spiffeid/TrustDomain.java:78]: `String.format()` is much slower than concatenation with `StringBuilder` or `+` (which the compiler will transform into using `StringBuilder`). I would rather see a concatenation used here since we don't really need any specific features of `String.format()`, i.e. decimal formatting

File: java-spiffe-core/src/main/java/io/spiffe/spiffeid/TrustDomain.java [hunk 2]
```diff
@@ -67,10 +69,23 @@ public String toString() {
...
+    }
+
     private static void validateHost(final String host) {
         if (StringUtils.isBlank(host)) {
             throw new IllegalArgumentException("Trust domain cannot be empty");
         }
+
+        if (host.length() > TRUST_DOMAIN_MAXIMUM_LENGTH) {
+            throw new IllegalArgumentException(String.format("Trust domain maximum length is %s bytes", TRUST_DOMAIN_MAXIMUM_LENGTH));
```
Comment (rturner3) [inline java-spiffe-core/src/main/java/io/spiffe/spiffeid/TrustDomain.java:87]: minor performance nit, the exception message can actually be a const field of this class since the format parameter is a const

File: java-spiffe-core/src/main/java/io/spiffe/spiffeid/SpiffeId.java [hunk 3]
```diff
@@ -86,13 +91,21 @@ public String toString() {
         return String.format("%s://%s%s", SPIFFE_SCHEME, this.trustDomain.toString(), this.path);
     }
 
-    private static String normalize(final String s) {
-        return s.toLowerCase().trim();
-    }
+    private static URI createAndValidateUri(final String uriAsString) {
+
```
Comment (rturner3) [inline java-spiffe-core/src/main/java/io/spiffe/spiffeid/SpiffeId.java:95]: nit: unnecessary whitespace

File: java-spiffe-core/src/main/java/io/spiffe/spiffeid/SpiffeId.java [hunk 4]
```diff
@@ -86,13 +91,21 @@ public String toString() {
...
+        URI uri;
+        try {
+            uri = URI.create(uriAsString.trim());
+        } catch (Exception e) {
+            throw new IllegalArgumentException(String.format("SPIFFE ID: malformed URI: %s", uriAsString), e);
+        }
+
+        if (uri.toASCIIString().length() > SPIFFE_ID_MAXIMUM_LENGTH) {
+            throw new IllegalArgumentException(String.format("SPIFFE ID: maximum length is %s bytes", SPIFFE_ID_MAXIMUM_LENGTH));
```
Comment (rturner3) [inline java-spiffe-core/src/main/java/io/spiffe/spiffeid/SpiffeId.java:104]: performance nit: this exception message can be a const field in this class

File: java-spiffe-core/src/test/java/io/spiffe/spiffeid/SpiffeIdTest.java [hunk 5]
```diff
@@ -1,89 +1,136 @@
...
 
-    @Test
-    void of_TrustDomainAndPathWithLeadingAndTrailingBlanks_ReturnsSpiffeIdNormalized() {
-        val trustDomain = TrustDomain.of(" trust-domain.org ");
-
-        val spiffeId = SpiffeId.of(trustDomain, " path1 ", " path2 ");
+    @ParameterizedTest
+    @MethodSource("provideValidTrustDomainAndPaths")
+    void testOf(TrustDomain inputTrustDomain, String[] inputPath, SpiffeId expectedSpiffeId) {
+        SpiffeId result;
```
Comment (rturner3) [inline java-spiffe-core/src/test/java/io/spiffe/spiffeid/SpiffeIdTest.java:89]: this can be combined with the initialization below since it's unused outside the `try`

File: java-spiffe-core/src/test/java/io/spiffe/spiffeid/SpiffeIdTest.java [hunk 6]
```diff
@@ -1,89 +1,136 @@
...
-        val trustDomain = TrustDomain.of("TRuST-DoMAIN.Org");
-
-        val spiffeId = SpiffeId.of(trustDomain, "PATH1", "paTH2");
+    @ParameterizedTest
+    @MethodSource("provideInvalidSpiffeIds")
+    void testParseInvalidSpiffeId(String input, String expected) {
+        SpiffeId result;
+        try {
+            result = SpiffeId.parse(input);
+            assertEquals(expected, result.toString());
```
Comment (rturner3) [inline java-spiffe-core/src/test/java/io/spiffe/spiffeid/SpiffeIdTest.java:63]: should this be a `fail()`? are there cases where parsing invalid spiffe ids don't result in an exception?

Comment (maxlambrecht) [reply java-spiffe-core/src/test/java/io/spiffe/spiffeid/SpiffeIdTest.java:63]: good catch!

**Review by rturner3 — APPROVED** [review]
